orchagent/portsorch.cpp

```diff
@@ -752,62 +752,200 @@ bool PortsOrch::setPortPfcAsym(Port &port, string pfc_asym)
+
+    attr.value.oid = bind ? (ingress ? port.m_ingress_acl_table_group_id :
+	                               port.m_egress_acl_table_group_id):
+			    SAI_NULL_OBJECT_ID;
+    switch (port.m_type)
+    {
+        case Port::PHY:
+        {
+            attr.id = ingress ?
+		    SAI_PORT_ATTR_INGRESS_ACL : SAI_PORT_ATTR_EGRESS_ACL;
+            status = sai_port_api->set_port_attribute(port.m_port_id, &attr);
+            break;
+        }
+        case Port::LAG:
+        {
+            attr.id = ingress ?
+		    SAI_LAG_ATTR_INGRESS_ACL : SAI_LAG_ATTR_EGRESS_ACL;
+            status = sai_lag_api->set_lag_attribute(port.m_lag_id, &attr);
+            break;
+        }
+        case Port::VLAN:
+        {
+            attr.id = ingress ?
+		    SAI_VLAN_ATTR_INGRESS_ACL : SAI_VLAN_ATTR_EGRESS_ACL;
+            status =
+	        sai_vlan_api->set_vlan_attribute(port.m_vlan_info.vlan_oid,
+				                 &attr);
+            break;
+        }
+        default:
+        {
+            SWSS_LOG_ERROR("Failed to %s %s port with type %d",
+			   bind_str.c_str(), port.m_alias.c_str(), port.m_type);
+            return false;
+        }
+    }
+
+    if (SAI_STATUS_SUCCESS != status)
+    {
+        SWSS_LOG_ERROR("Failed to %s %s to ACL table group %" PRIx64 ", rv:%d",
+                       bind_str.c_str(), port.m_alias.c_str(), attr.value.oid, status);
+        return false;
+    }
+
+    return true;
+}
+
+bool PortsOrch::unbindRemoveAclTableGroup(sai_object_id_t  port_oid,
+		                          sai_object_id_t  acl_table_oid,
+					  acl_stage_type_t acl_stage)
 {
     SWSS_LOG_ENTER();
 
-    if (acl_stage == ACL_STAGE_UNKNOWN)
+    sai_status_t       status;
+    bool               ingress = (acl_stage == ACL_STAGE_INGRESS);
+    Port               port;
+
+    if (!getPort(port_oid, port))
     {
-        SWSS_LOG_ERROR("unknown ACL stage for table group creation");
+        SWSS_LOG_ERROR("Failed to get port by port OID %" PRIx64, port_oid);
         return false;
     }
 
-    Port port;
-    if (!getPort(id, port))
+
+    sai_object_id_t &group_oid_ref = 
+	    ingress? port.m_ingress_acl_table_group_id :
+	             port.m_egress_acl_table_group_id;
+    unordered_set<sai_object_id_t> &acl_list_ref = 
+	    ingress ? port.m_ingress_acl_tables_uset :
+	              port.m_egress_acl_tables_uset;
+
+    if (SAI_NULL_OBJECT_ID == group_oid_ref)
+    {
+        assert(acl_list_ref.find(acl_table_oid) == acl_list_ref.end());
+        return true;
+    }
+    assert(acl_list_ref.find(acl_table_oid) != acl_list_ref.end());
+    acl_list_ref.erase(acl_table_oid);
+    if (!acl_list_ref.empty()) 
+    {
+	// This port is in more than one acl table's port list
+	// So, we need to preserve group OID
+        SWSS_LOG_NOTICE("Preserving port OID %" PRIx64" ACL table grop ID", port_oid);
+        setPort(port.m_alias, port);
+        return true;
+    }
+
+    port.m_dependency_bitmap &= ~(1 << Port::ACL_DEP);
```

Comment:
make this as  a member function so we could easily use different approaches and hide the internals.

Reply:
Makes sense. Done!